lib/spack/spack/build_systems/cuda.py

```diff
@@ -0,0 +1,73 @@
+##############################################################################
+# Copyright (c) 2013-2017, Lawrence Livermore National Security, LLC.
+# Produced at the Lawrence Livermore National Laboratory.
+#
+# This file is part of Spack.
+# Created by Todd Gamblin, tgamblin@llnl.gov, All rights reserved.
+# LLNL-CODE-647188
+#
+# For details, see https://github.com/llnl/spack
+# Please also see the NOTICE and LICENSE files for our notice and the LGPL.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+import os
+from spack import *
+
+
+class CudaPackage(Package):
+    """Auxiliary class which contains CUDA variant, dependencies and conflicts
+    and is meant to unify and facilitate its usage.
+    """
+
+    # FIXME: keep cuda and cuda_arch separate to make usage easier untill
+    # Spack has depends_on(cuda, when='cuda_arch!=None') or alike
+    variant('cuda', default=False,
+            description='Build with CUDA')
+    # see http://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#gpu-feature-list
+    variant('cuda_arch', default=None,
+            description='CUDA architecture',
+            values=('20', '30', '32', '35', '50', '52', '53', '60', '61', '62', '70'),
+            multi=False)
+    # see http://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-examples
+    # and http://llvm.org/docs/CompileCudaWithLLVM.html#compiling-cuda-code
+    @property
+    def cuda_flags(self, arch_list):
+        return ['--generate-code arch=compute_{0},code=sm_{0}'.format(s) for s in arch_list]
+
+    depends_on("cuda@7:", when='+cuda')
+
+    # CUDA version vs Architecture
+    depends_on("cuda@8:", when='cuda_arch=60')
+    depends_on("cuda@8:", when='cuda_arch=61')
+    depends_on("cuda@8:", when='cuda_arch=62')
+    depends_on("cuda@9:", when='cuda_arch=70')
+
+    depends_on('cuda@:8.99', when='cuda_arch=20')
+
+    # Compiler conflicts:
+    conflicts('+cuda', when='%gcc@6:')
```

Comment:
cuda 9 supports gcc 6 and lower, cuda 8 supports gcc 5 and lower, cuda 7.5 supports gcc 4.9 and lower, etc.